### ISSUE TITLE
Expected values from ShapValues do not match with prediction

### ISSUE THREAD

**Issue by ibuda** [catboost/catboost#1170]
This is a recurrent issue already mentioned in #1139 
I have updated catboost to the latest and greatest, which is currently the 0.21 version, but the issue still persists, i.e. the prediction value does not correspond to expected_value taken from ShapValues. Please see the code below to reproduce the issue, or #1139 for more details:
```
import catboost
from catboost import CatBoostClassifier, Pool
from catboost.datasets import titanic
from sklearn.model_selection import train_test_split
import numpy as np

train_df, test_df = catboost.datasets.titanic()
train_df.fillna(-999, inplace=True)
test_df.fillna(-999, inplace=True)
X = train_df.drop('Survived', axis=1)
y = train_df.Survived

cat_ix = np.where(X.dtypes != np.float)[0]
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=42)
X_test = test_df
params = {
    'iterations': 500,
    'learning_rate': 0.1,
    'eval_metric': 'Accuracy',
    'random_seed': 42,
    'logging_level': 'Silent',
}
train_pool = Pool(X_train, y_train, cat_features=cat_ix)
validate_pool = Pool(X_val, y_val, cat_features=cat_ix)
model = CatBoostClassifier(**params)
model.fit(train_pool, eval_set=validate_pool)

print(model.predict(train_pool, prediction_type='RawFormulaVal').mean().round(4))
# prints -0.7318

shap_values = model.get_feature_importance(train_pool, type="ShapValues")
expected_value = shap_values[:,-1][0]
print(expected_value.round(4))
# prints -0.5393
print(catboost.__version__)
# prints '0.21'
# shap_values.sum(axis=1).mean().round(4) # -0.7318
```

 I know that the last release should've fixed this, but unfortunately, the issue persists. In case I am wrong, I apologize and gratefully ask you to point out on how to bypass this issue besides the `shap_values.sum(axis=1).mean().round(4)` workaround.

Problem: Expected values from ShapValues do not match with prediction
catboost version: 0.21
Operating System: Ubuntu 19.04
CPU: Intel® Core™ i5-7400 CPU

**Comment by LSinev** [comment]
@ibuda Which version was the latest not affected by that and previous issue? 
Is TreeExplainer from `shap` package affected?
I am using  `model.get_feature_importance(train_pool, type="ShapValues")` for feature selection, but if  is not correct, should i prefer  `TreeExplainer` from `shap`?

**Comment by ibuda** [comment]
> Which version was the latest not affected by that and previous issue?

As a matter of fact, up to my knowledge, this issue has been there from the very start.
> Is TreeExplainer from shap package affected?

I do not know what exactly you mean here, but if you're interested in the discussion on this matter, you can see this [issue](https://github.com/slundberg/shap/issues/966) and the temporary workaround I proposed there for that very issue.
> should i prefer TreeExplainer from shap?

works like a charm for me. However, I think that catboost's methods also should work, as this inconsistency does not affect the "rank" or "importance" of the model, just the last expected value in the array. 

In one word, this entire topic is not a major bug, but a small discrepancy, which I think will be quickly "fixed" by catboost team.

**Comment by annaveronika** [comment]
The difference comes from the way, how we calculate, which leaf a training sample belongs to.
If you have no categorical features, then if you calculate it during training or after training - it will be the same.
If you have categorical features, then there might be a difference, because new numerical features are calculated differently during training and after training:
https://catboost.ai/docs/concepts/faq.html#applying-the-model-to-train-dataset

So the reason for the difference is that predictions for training samples that you do when you call model.predict() are not using features that are calculated on a random permutation, but are using features that are calculated on the whole training dataset.

So there is no bug, this is just the way numerical features are calculated based on categorical ones for the training dataset.

**Comment by ibuda** [comment]
@annaveronika Thank you for stepping in and explaining this matter for us. As I mentioned above, I did not say it is a bug, but a discrepancy, which needs clarification. Once again, thank you for a great product and for clearing this issue out for us.
Closing the issue.
